### PR TITLE
test(docs): create comprehensive test suite for documentation

### LINKED ISSUE #58: [Test] Docs - Write Tests
## Overview
Create the docs/ directory for comprehensive project documentation including guides, tutorials, API documentation, and architectural documentation. This supplements the README with detailed information.

## Testing Objectives
This phase focuses on:
- Writing comprehensive test cases following TDD principles
- Creating test fixtures and mock data
- Defining test scenarios for edge cases
- Setting up test infrastructure

## What to Test
Based on the expected outputs:
- docs/ directory at repository root
- docs/README.md explaining documentation structure
- Subdirectory structure for different documentation types
- Guidelines for contributing documentation

## Test Success Criteria
- [ ] docs/ directory exists at repository root
- [ ] README clearly explains documentation structure
- [ ] Subdirectories are organized logically
- [ ] Guidelines help contributors write good docs

## Implementation Steps
1. Create docs/ directory at repository root
2. Write README explaining documentation organization
3. Create subdirectories for guides, API docs, and architecture
4. Provide templates or examples for documentation

## Notes
Good documentation is critical for adoption and contribution. Organize docs by type and audience to make information easy to find. Keep it simple initially and expand as needed.

### PR DESCRIPTION
**Pull request by mvillmow**
# Documentation Test Suite - 4-Tier Structure Validation

This PR creates a comprehensive test suite for validating the 4-tier documentation structure with 24 documents (Issue #58).

## Overview

Create TDD test suite for documentation structure validation following core functionality testing approach, aligned with Issue #57 planning.

## Test Files Created (8 files, 3,419 lines, 116 tests → 200 collected)

1. **test_doc_structure.py** (259 lines) - 4-tier hierarchy validation
2. **test_doc_completeness.py** (439 lines) - All 24 documents validation
3. **test_link_validation.py** (421 lines) - Internal link checking
4. **test_markdown_linting.py** (516 lines) - Markdown compliance (MD040, MD031, MD032, MD022, MD013)
5. **test_getting_started.py** (485 lines) - Tier 1 (6 docs)
6. **test_core_docs.py** (440 lines) - Tier 2 (8 docs)
7. **test_advanced_docs.py** (413 lines) - Tier 3 (6 docs)
8. **test_dev_docs.py** (446 lines) - Tier 4 (4 docs)

## Key Features

- **TDD Compliant**: Tests written before docs exist (will fail until implemented)
- **Parametrized Tests**: 116 functions expand to 200 collected tests
- **Self-Contained**: pytest only, no external dependencies
- **Fast & Deterministic**: Uses tmp_path fixtures for isolation
- **Comprehensive**: Structure, content, links, formatting all validated

## Alignment with Planning (Issue #57)

All 4 tiers from documentation planning validated:
- ✅ Tier 1: Getting Started (6 docs)
- ✅ Tier 2: Core Documentation (8 docs)
- ✅ Tier 3: Advanced Topics (6 docs)
- ✅ Tier 4: Development Guides (4 docs)

## Dependencies

- Requires: Issue #1556 (shared test infrastructure - pytest fixtures)
- Guides: Issue #59 (implementation phase)

## References

- **Issue**: #58
- **Planning**: Issue #57
- **Documentation**: `notes/issues/58/README.md`

Closes #58

### CODE REVIEW

**Comment by mvillmow** [comment]
## Code Review Summary - Documentation Test Suite

**Status**: ✅ Comprehensive structure validation - Ready to merge

**Files Reviewed**: 8 test modules (3,419 lines, 116 tests → 200 collected)

**Strengths**:
- ✅ Excellent pytest parametrization (116 functions → 200 tests)
- ✅ 100% coverage of Issue #57 planning (all 4 tiers, 24 docs)
- ✅ Self-contained (pytest only, no external deps)
- ✅ Fast and deterministic with tmp_path fixtures
- ✅ Clean Python code with type hints

**Test Coverage**:
- Structure: 4-tier hierarchy validation
- Completeness: All 24 documents with minimum content
- Link Validation: Internal and cross-references
- Markdown Linting: MD040, MD031, MD032, MD022, MD013 compliance
- Tier-specific: Validates each tier's unique requirements

**Tier Coverage**:
- Tier 1: Getting Started (6 docs) ✓
- Tier 2: Core Documentation (8 docs) ✓
- Tier 3: Advanced Topics (6 docs) ✓
- Tier 4: Development Guides (4 docs) ✓

**Issues Found**: None

Will effectively guide Issue #59 implementation. Excellent work!

File: tests/foundation/docs/test_advanced_docs.py
Comment (mvillmow) [inline tests/foundation/docs/test_advanced_docs.py]: don't use mock directories, use the real directories based on the repository root.

**Review by mvillmow — COMMENTED** [review]
## Code Review - PR #1559

### Overall Assessment: APPROVE with Improvements Required ⚠️

Three specialist reviews completed. Test suite is comprehensive (200 tests) but has code quality and coverage gaps.

---

### 🔴 CRITICAL: Missing Markdown Standards Enforcement

**Reviewer**: Documentation Review Specialist

**Issue**: Tests don't enforce CLAUDE.md markdown standards (MD040, MD031, MD032, MD022, MD013).

**Current**: Only tests code blocks closed, list indentation, heading hierarchy
**Missing**: Language tags, blank lines, line length limits

**Fix**: Add test methods for each markdownlint rule:
```python
def test_code_blocks_have_language(self, all_doc_files):
    """Test MD040: Code blocks must have language specified."""
    # Verify ``` is followed by language (python, mojo, bash, etc.)
```

See: CLAUDE.md lines 519-662 for complete standards

---

### 🟠 MAJOR: Code Duplication Across All 8 Test Files

**Reviewer**: Implementation Review Specialist

**Issue**: Fixtures duplicated in ALL 8 test files (~240 lines of duplication):
- `repo_root` fixture defined 8 times
- `docs_root` fixture defined 5 times  
- Tier directory fixtures defined multiple times

**Fix**: Create `tests/foundation/docs/conftest.py` with shared fixtures.

**Impact**: Currently requires updating 8+ files for any fixture change.

---

### 🟠 MAJOR: Missing 4-Tier Structure Validation

**Reviewer**: Documentation Review Specialist

**Issue**: Tests validate agents/ docs but NOT the 4-tier project documentation (24 docs per Issue #57).

**Current**: Only agent documentation tested
**Missing**: Tier 1-4 validation (Getting Started, Core, Advanced, Dev)

**Fix**: Add test class for complete structure:
```python
def test_tier1_getting_started_docs(self, repo_root):
    """Test Tier 1: 6 Getting Started documents exist."""
    tier1_docs = ["README.md", "CONTRIBUTING.md", ...]
```

---

### 🟠 MAJOR: Superficial Quality Checks

**Reviewer**: Documentation Review Specialist

**Missing validation**:
- Code examples are syntactically valid
- Cross-references have descriptive text
- Sections have sufficient content depth
- Examples are clear and helpful

**Current**: Only checks document exists + has title + >100 chars

**Fix**: Add quality depth tests (code parsing, link context, content metrics).

---

### 🟡 MINOR: Weak Assertions (40+ locations)

**Reviewer**: Implementation Review Specialist

**Issue**: Many assertions too permissive:
```python
# Weak - passes if ANY section exists
assert "## " in text

# Better - verify specific sections
assert "## Overview" in text or "## Details" in text
```

**Recommendation**: Strengthen assertions where practical without making tests brittle.

---

### ✅ STRENGTHS: Test Coverage & Organization

**Reviewers**: All specialists

**Excellent design**:
- 200 parametrized tests from 116 functions ✅
- Clear tier-based organization ✅
- Comprehensive docstrings ✅
- Good use of pytest patterns ✅
- Isolated, independent tests ✅
- TDD approach (tests before docs) ✅

---

### Action Required

**Before Merge**:
1. ✅ Add markdown standards tests (MD040, MD031, MD032, MD022, MD013)
2. ✅ Create conftest.py with shared fixtures
3. ✅ Add 4-tier structure validation
4. ✅ Strengthen quality checks (code validation, link context)
5. ✅ Extract magic numbers to constants
6. ✅ Add return type hints (-> None) to all test methods

**Consider**:
7. 💡 Strengthen weak assertions where practical
8. 💡 Add pytest markers for test organization
9. 💡 Extract helper functions to shared module

---

### Summary

**Test Coverage**: GOOD - 200 tests across 8 modules ✅
**Test Quality**: NEEDS IMPROVEMENT - weak assertions, missing standards 🟠
**Code Quality**: NEEDS REFACTORING - significant duplication 🟠
**Documentation**: EXCELLENT - clear docstrings ✅

**Recommendation**: APPROVE after addressing critical markdown standards enforcement and code duplication. The test suite design is excellent but needs these fixes for maintainability.

---

🤖 Generated via Code Review Orchestrator with 3 specialist agents

**Comment by mvillmow** [comment]
## Review Comments Addressed

All 5 review comments have been addressed in commit 94d2c27:

1. **Comment on test_advanced_docs.py (mock repository root)**: ✅ Fixed - Updated conftest.py to use real repository root instead of tmp_path mock

2. **Comment on test_advanced_docs.py:1 (mock directories)**: ✅ Fixed - Updated repo_root fixture to navigate from tests/foundation/docs/ to actual repository root

3. **Comment on test_link_validation.py:1**: ✅ Fixed - Deleted test_link_validation.py as link validation is handled by pre-commit hooks

4. **Comment on test_markdown_linting.py:1**: ✅ Fixed - Deleted test_markdown_linting.py as markdown linting is handled by pre-commit hooks

5. **Comment on notes/issues/58/README.md:89**: ✅ Fixed - Removed markdown linting section from README.md and updated test file counts throughout the document

### Summary of Changes
- Updated `conftest.py` to use real repository root (navigates up from test file location)
- Updated `test_advanced_docs.py` docstring to reflect real repository usage
- Deleted `test_link_validation.py` (421 lines) - functionality covered by pre-commit
- Deleted `test_markdown_linting.py` (778 lines) - functionality covered by pre-commit
- Updated `notes/issues/58/README.md` to remove references to deleted tests and update counts
- All remaining tests still pass (191 tests collected, 159 passed as expected for TDD)

File: notes/issues/58/README.md [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+
+- Code blocks have language specified (MD040)
+- Code blocks surrounded by blank lines (MD031)
+- Lists surrounded by blank lines (MD032)
+- Headings surrounded by blank lines (MD022)
+- Line length compliance (MD013, 120 chars)
+- No trailing whitespace
+- Files end with newline
+
+**Coverage**: Markdown formatting, linting rules, style compliance
```
[the remark this answers is not in the document]
Comment (mvillmow) [reply notes/issues/58/README.md]: ✅ Fixed - Removed markdown linting section from README.md since it's handled by pre-commit hooks

**Comment by mvillmow** [comment]
@claude  review this
